### PR TITLE
add support for absolute hydration

### PR DESCRIPTION
**Pull request by jkausrelay**
- Follows rbfe so we can reuse estimate_free_energy_given_initial_states
- add a simple test

### CODE REVIEW

File: timemachine/fe/absolute_hydration.py [hunk 1]
```diff
@@ -131,6 +148,7 @@ def set_up_ahfe_system_for_smc(
     mol, n_walkers, n_windows, n_md_steps, resample_thresh, seed=2022, ff=None, num_workers=None
 ):
     """define initial samples, lambdas schedule, propagate fxn, log_prob fxn, resample fxn"""
+    n_windows = n_windows or 16
```
Comment (jkausrelay) [inline timemachine/fe/absolute_hydration.py:151]: 16 is arbitrary here, need to do some tests w/free solv to determine a good default

File: timemachine/fe/absolute_hydration.py [hunk 2]
```diff
@@ -155,3 +173,184 @@ def log_prob(xs, lam):
...
+) -> List[InitialState]:
+    """
+    Setup the initial states for a series of lambda values. It is assumed that the lambda schedule
+    is a monotonically increasing sequence in the closed interval [0, 1].
+
+    Parameters
+    ----------
+    afe: AbsoluteFreeEnergy
+        An AbsoluteFreeEnergy object which contains the mol structure
+
```
Comment (maxentile) [inline timemachine/fe/absolute_hydration.py:274]: nit[completeness]: missing parameter in docstring

File: timemachine/fe/absolute_hydration.py [hunk 3]
```diff
@@ -155,3 +173,184 @@ def log_prob(xs, lam):
...
+    host_config: HostConfig
+        Configurations of the host.
+
+    temperature: float
+        Temperature to run the simulation at.
+
+    lambda_schedule: list of float of length K
+        Lambda schedule.
```
Comment (maxentile) [inline timemachine/fe/absolute_hydration.py:282]: nit[style opinion] -- parameter descriptions of this form can be omitted.

Consider either removal or expansion. (E.g. "must start at 1 and end at 0")

File: timemachine/fe/rbfe.py [hunk 4]
```diff
@@ -178,8 +178,8 @@ def setup_initial_states(st, host_config, temperature, lambda_schedule, seed):
 
     initial_states = []
 
-    # check that the lambda schedule is monotonically increasing.
-    assert np.all(np.diff(lambda_schedule) > 0)
+    # check that the lambda schedule is monotonically increasing or decreasing.
+    assert np.all(np.diff(lambda_schedule) > 0) or np.all(np.diff(lambda_schedule) < 0)
```
Comment (maxentile) [inline timemachine/fe/rbfe.py:182]: Possible bug: This looks like it weakens an assertion when used for RBFE (unless RBFE estimate is already invariant to reversal of lambda schedule).

(Consider removal, or specializing to assert one order for RBFE and the opposite order for AHFE, or re-introducing the original assertion in a still-RBFE-specific codepath.)

**Review by maxentile — APPROVED** [review]

File: timemachine/fe/rbfe.py (hunk 4 above)
Comment (jkausrelay) [reply timemachine/fe/rbfe.py:182]: Definitely not intended, reverting.

File: timemachine/fe/absolute_hydration.py [hunk 5]
```diff
@@ -155,3 +173,184 @@ def log_prob(xs, lam):
...
+    afe: AbsoluteFreeEnergy,
+    ff: Forcefield,
+    host_config: HostConfig,
+    temperature: float,
+    lambda_schedule: Array,
+    seed: int,
+) -> List[InitialState]:
+    """
+    Setup the initial states for a series of lambda values. It is assumed that the lambda schedule
+    is a monotonically increasing sequence in the closed interval [0, 1].
```
Comment (mcwitt) [inline timemachine/fe/absolute_hydration.py:267]: Should this say "decreasing"? (we pass in a decreasing sequence on L240)

File: timemachine/fe/absolute_hydration.py [hunk 6]
```diff
@@ -155,3 +172,186 @@ def log_prob(xs, lam):
...
+    SimulationResult
+        Collected data from the simulation (see class for storage information). We currently return frames
+        from only the first and last window.
+    """
+    bt = BaseTopology(mol, ff)
+    afe = AbsoluteFreeEnergy(mol, bt)
+
+    # note: tm convention lambda=1 means "decoupled", lambda=0 means "coupled"
+    lambda_schedule = construct_pre_optimized_absolute_lambda_schedule_solvent(n_windows)[::-1]
+    assert np.isclose(lambda_schedule[0], 1.0) and np.isclose(lambda_schedule[-1], 0.0)
```
Comment (mcwitt) [inline timemachine/fe/absolute_hydration.py:236]: nit: any reason not to assert equality here?

File: timemachine/fe/absolute_hydration.py (hunk 5 above)
Comment (jkausrelay) [reply timemachine/fe/absolute_hydration.py:267]: yes

File: timemachine/fe/absolute_hydration.py (hunk 1 above)
Comment (jkausrelay) [reply timemachine/fe/absolute_hydration.py:151]: No, moved to run_solvent

**Comment by jkausrelay** [comment]
- I moved the common dataclasses to fe.free_energy
- Left estimate_free_energy_given_initial_states in rbfe since that requires a larger refactor

**Review by mcwitt — APPROVED** [review]

File: timemachine/fe/absolute_hydration.py (hunk 6 above)
Comment (jkausrelay) [reply timemachine/fe/absolute_hydration.py:233]: just used the same code as above

**Pull request merged**